java-cfenv-jdbc/src/main/java/io/pivotal/cfenv/jdbc/MySqlJdbcUrlCreator.java

```diff
@@ -23,48 +23,72 @@
+                || cfService.existsByCredentialsContainsUriField(MYSQL_SCHEMES);
+    }
 
-	public static final String MYSQL_TAG = "mysql";
+    private boolean isMariaDbService(CfService cfService) {
+        return jdbcUrlMatchesScheme(cfService, MARIADB_SCHEME)
+                || cfService.existsByTagIgnoreCase(MARIADB_TAG)
+                || cfService.existsByLabelStartsWith(MARIADB_LABEL)
+                || cfService.existsByUriSchemeStartsWith(MARIADB_SCHEME)
+                || cfService.existsByCredentialsContainsUriField(MARIADB_SCHEME);
+    }
 
-	public static final String MYSQL_LABEL = "mysql";
+    @Override
+    public String getDriverClassName() {
+        String driverClassNameToUse;
+        try {
+            driverClassNameToUse = MARIADB_DRIVER_CLASS_NAME;
+            Class.forName(driverClassNameToUse, false, getClass().getClassLoader());
+        }
+        catch (ClassNotFoundException e) {
+            try {
+                driverClassNameToUse = MYSQL_DRIVER_CLASS_NAME;
+                Class.forName(driverClassNameToUse, false, getClass().getClassLoader());
+            }
+            catch (ClassNotFoundException e2) {
+                return null;
+            }
+        }
+        return driverClassNameToUse;
+    }
 
-	@Override
-	public boolean isDatabaseService(CfService cfService) {
-		if (jdbcUrlMatchesScheme(cfService, MYSQL_SCHEME)
-				|| cfService.existsByTagIgnoreCase(MYSQL_TAG)
-				|| cfService.existsByLabelStartsWith(MYSQL_LABEL)
-				|| cfService.existsByUriSchemeStartsWith(MYSQL_SCHEME)
-				|| cfService.existsByCredentialsContainsUriField(MYSQL_SCHEME)) {
-			return true;
-		}
-		return false;
-	}
-
-	@Override
-	public String getDriverClassName() {
-		String driverClassNameToUse = null;
-		try {
-			driverClassNameToUse = "org.mariadb.jdbc.Driver";
-			Class.forName(driverClassNameToUse, false, getClass().getClassLoader());
-		} catch (ClassNotFoundException e) {
-			try {
-				driverClassNameToUse = "com.mysql.cj.jdbc.Driver";
-				Class.forName(driverClassNameToUse, false, getClass().getClassLoader());
-			} catch (ClassNotFoundException e2) {
-				return null;
-			}
-		}
-		return driverClassNameToUse;
-	}
-
-	@Override
-	public String buildJdbcUrlFromUriField(CfCredentials cfCredentials) {
-		UriInfo uriInfo = cfCredentials.getUriInfo(MYSQL_SCHEME);
-		return String.format("%s%s://%s%s/%s%s%s", JDBC_PREFIX, MYSQL_SCHEME,
-				uriInfo.getHost(), uriInfo.formatPort(), uriInfo.getPath(),
-				uriInfo.formatUserNameAndPasswordQuery(), uriInfo.formatQuery());
-	}
+    @Override
+    public String createJdbcUrl(CfService cfService) {
+        CfCredentials cfCredentials = cfService.getCredentials();
+        String jdbcUrl = (String) cfCredentials.getMap().get("jdbcUrl");
+        if (jdbcUrl == null) {
+            jdbcUrl = buildJdbcUrlFromUriField(cfCredentials);
+        }
+        if (isMariaDbService(cfService) && getDriverClassName().equals(MARIADB_DRIVER_CLASS_NAME)) {
```

Comment:
Hello @anthonydahanne and thanks for the response.

The reason I didn't use buildJdbcUrlFromUriField(CfCredentials cfCredentials) and decided to override createJdbcUrl(CfService cfService) was because buildJdbcUrlFromUriField would only resolve the issue insofar as _creating_ JDBC urls with MariaDB prefixes if the jdbcUrl isn't already present in the credentials of the VCAP_SERVICES environment variable. A case that was encountered and justified pursuing this avenue, was a MariaDB CF service where the jdbcUrl would already be set with a mysql prefix, and the library would inject the jdbcUrl (with Spring Boot) as-is.

The other reason why I didn't use that method was because it lacks the CfService cfService argument, so I couldn't make use of the isMariaDbService(CfService cfService) function to additionally determine if the current service is a MariaDB service, being restricted to determining if the service is a MariaDB or MySQL service based solely on the driver class name, which would conflict with most existing test cases, and with applications that use MySQL database servers with the MariaDB Connector/J.

These reasons warranted for the JDBC url being retrieved or built, and then depending on the MariaDB criteria and the current database driver being patched with the MariaDB JDBC prefix in the MySqlJdbcUrlCreator.

I will replace the use of spaces by tabs in the updated files.